### PR TITLE
[#4615] Customizable estimation for messages written outside the EventLoop

### PR DESCRIPTION
**Pull request by technocoreai**
Motivation:

Estimation algorithm currently used for WriteTasks is complicated and wrong. Additionally, some code relies on outbound buffer size incremented only on actual writes to the outbound buffer.

Modifications:

- Throw away the old estimator and replace with a simple algorithm that uses the client-provided estimator along with a statically configured WriteTask overhead (`io.netty.transport.writeTaskSizeOverhead` system property with the default value of 48 bytes)
- Add a `io.netty.transport.estimateSizeOnSubmit` boolean system property allowing the clients to disable the message estimation outside the event loop

Result:

Task estimation is user controllable and produces better results bydefault

### CODE REVIEW

**Comment by technocoreai** [comment]
This is for the 4.1 branch, I'll cross-port to 4.0/5.0 once you approve. Default task size is hardcoded, I could do something more fancy like detecting the JVM bitness and `CompressedOOPS`, but it's way too complicated and the users can just change the value as necessary.

File: transport/src/main/java/io/netty/channel/DefaultChannelHandlerInvoker.java [hunk 1]
```diff
@@ -506,13 +419,20 @@ private static WriteTask newInstance(
...
             task.promise = promise;
-            task.size = ((AbstractChannel) ctx.channel()).estimatorHandle().size(msg)
-                    + estimateSize(task, CLASS_SIZES.get());
-            ChannelOutboundBuffer buffer = ctx.channel().unsafe().outboundBuffer();
-            // Check for null as it may be set to null if the channel is closed already
-            if (buffer != null) {
-                buffer.incrementPendingOutboundBytes(task.size);
+
+            if (ESTIMATE_TASK_SIZE_ON_SUBMIT) {
+                task.size = ((AbstractChannel) ctx.channel()).estimatorHandle().size(msg) + WRITE_TASK_OVERHEAD;
```
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/DefaultChannelHandlerInvoker.java]: @technocoreai just one nit.. can you move this in the `if (buffer != null) { ... }` block and if the buffer == null set the size = 0 ?

**Comment by normanmaurer** [comment]
@technocoreai only one tiny comment. No need to port this to other branches, I will just cherry-pick it from 4.1 once in there.

**Comment by technocoreai** [comment]
@normanmaurer I've updated the patch, it now initializes the size to 0 and then updates it if the estimation is enabled and channel is still open, is that fine?

File: transport/src/main/java/io/netty/channel/DefaultChannelHandlerInvoker.java [hunk 2]
```diff
@@ -506,13 +419,18 @@ private static WriteTask newInstance(
             task.ctx = ctx;
             task.msg = msg;
             task.promise = promise;
-            task.size = ((AbstractChannel) ctx.channel()).estimatorHandle().size(msg)
-                    + estimateSize(task, CLASS_SIZES.get());
-            ChannelOutboundBuffer buffer = ctx.channel().unsafe().outboundBuffer();
-            // Check for null as it may be set to null if the channel is closed already
-            if (buffer != null) {
-                buffer.incrementPendingOutboundBytes(task.size);
+            task.size = 0;
```
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/DefaultChannelHandlerInvoker.java]: not need to set this to 0 if ESTIMATE_TASK_SIZE_ON_SUBMIT is true. move this to an else block please.

Comment (technocoreai) [inline transport/src/main/java/io/netty/channel/DefaultChannelHandlerInvoker.java]: @normanmaurer Fixed, although I don't think that the size of a task written to a closed channel matters much.

**Comment by normanmaurer** [comment]
Cherry-picked into 4.0 (cfd6793bb7b1ae6769cf0b3eaffd1aecfa6fbd32) and 4.1 (cfd6793bb7b1ae6769cf0b3eaffd1aecfa6fbd32)

@technocoreai thanks again!

**Comment by technocoreai** [comment]
@normanmaurer The commit looks the same for branches and I don't see it in the log. Do you intend to push the 4.0 fix later?

**Comment by normanmaurer** [comment]
@technocoreai sorry I somehow thought I pushed to 4.0 as well. Just updated the comment to link to the correct commit sha and pushed it. d2ddb528e4427d201866280017176723cfb93a5a

**Comment by technocoreai** [comment]
@normanmaurer thanks!

**Pull request closed without merging**